File: 40. Combination Sum II.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        self.res = []
        candidates = sorted(candidates)
        self.DFS([], 0, candidates, target)
        return self.res
        
    def DFS(self, path, start, candidates, target):
        if target == 0:
            self.res += [path]
            return
        i = start
        while i < len(candidates):
            if candidates[i] <= target:
                self.DFS(path + [candidates[i]], i+1, candidates, target - candidates[i])
            i += 1
            while i < len(candidates) and candidates[i] == candidates[i-1]:
                i += 1

File: 40. Combination Sum II.cpp
class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        DFS({}, 0, candidates, target);
        return res;
    }
    
    void DFS(vector <int> path, int start, vector<int>& candidates, int target)
    {
        if (target == 0)
        {
            res.push_back(path);
            return;
        }
        int i = start;
        while (i < candidates.size())
        {
            if (candidates[i] <= target)
            {
                path.push_back(candidates[i]);
                DFS(path, i + 1, candidates, target - candidates[i]);
                path.pop_back();
            }
            i++;
            while (i < candidates.size() && candidates[i] == candidates[i-1])
                i++;
        }
    }
};
